zephyr/syscall/sof_dma.c

```diff
@@ -0,0 +1,224 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation.
+
+#include <sof/lib/dma.h>
+#include <zephyr/kernel.h>
+#include <zephyr/internal/syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+
+static inline bool sof_dma_has_access(struct sof_dma *dma)
+{
+	/*
+	 * use the Zephyr dma.h device handle to check calling
+	 * thread has access to it
+	 */
+	return k_object_is_valid(dma->z_dev, K_OBJ_ANY);
+}
+
+static inline bool sof_dma_is_valid(struct sof_dma *dma)
+{
+	const struct dma_info *info = dma_info_get();
+	struct sof_dma *array_end = info->dma_array + info->num_dmas, *d;
+
+	if (!info->num_dmas)
+		return false;
+
+	for (d = info->dma_array; d < array_end; d++)
+		if (d == dma && sof_dma_has_access(dma))
+			return true;
```

Comment:
aah, sorry, it's *almost* good. This works, but it continues looping needlessly if you've found `d == dma` to which the user has no access. The best version should be
```
	if (d == dma)
		return sof_dma_has_access(dma);
```
but since CI cycles are precious we can do this as a follow-up. Unless you have to update for some other reason.